src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,188 @@
+package org.apache.spark.mllib.linalg.distributed
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.internal.Logging
+
+object BetterBlockMatrix {
+
+  class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix, deps: Seq[Dependency[_]], partitions: Array[Partition])
+      extends RDD[((Int, Int), Matrix)](l.blocks.sparkContext, deps) {
+
+    val lPartitioner = l.blocks.partitioner.get.asInstanceOf[GridPartitioner]
+    val rPartitioner = r.blocks.partitioner.get.asInstanceOf[GridPartitioner]
+    val nRows: Long = l.numRows()
+    val nCols: Long = r.numCols()
+    val rowsPerBlock: Int = l.rowsPerBlock
+    val colsPerBlock: Int = r.colsPerBlock
+    val rowBlocks = l.numRowBlocks
+    val colBlocks = r.numColBlocks
+    val nProducts = l.numColBlocks
+    val rowsRemainder: Int = (nRows % rowsPerBlock).toInt
+    val colsRemainder: Int = (nCols % colsPerBlock).toInt
+
+    private def toDenseMatrix(x: Matrix): DenseMatrix = x match {
+      case x: DenseMatrix => x
+      case x: SparseMatrix => x.toDense
+    }
+
+    private def multiplyAccumulateWithExtension(x: DenseMatrix, y: DenseMatrix, result: DenseMatrix) {
+      val x2 = if (result.numRows != x.numRows)
```

Comment:
The result is also truncated in size.  Everything matches up.  I still don't see it is necessary to pad.

Let's say those are NxK and KxM matrices and, moreover, the subdimensions are (n1, n2)x(k1, k2) and (k1, k2)x(m1, m2).  Then B times C' (the block below A') should be a n1xk2 * k2xm1 = n1xm1 multiply and that has the same shape as A.